src/modules/RA/DataLayer/components/FieldUpdater.js

```diff
@@ -0,0 +1,115 @@
+import { useEffect } from 'react';
+import get from 'lodash.get';
+/* eslint-disable import/no-extraneous-dependencies */
+import { connect } from 'react-redux';
+import { change } from 'redux-form';
+/* eslint-enable */
+import {
+  withDataProvider,
+  GET_ONE,
+  REDUX_FORM_NAME,
+} from 'react-admin';
+
+import { WMTS } from '../../DataSource';
+import { RES_DATASOURCE } from '../../ra-modules';
+import compose from '../../../../utils/compose';
+
+
+/** Merge layer fields and source fields by preserving
+ * layer fields ordering and adding source information.
+ * This is a tree step merge:
+ * - Remove layer fields that are not in source fields
+ * - Update layer fields with source fields
+ * - Add new source fields
+ */
+const mergeFields = (layerFields, sourceFields) => {
+  // Keep fields from layer that are actually in source
+  const keptLayerFields = layerFields.filter(layerField =>
+    sourceFields.some(sourceField => layerField.field === sourceField.field));
+
+  // Update existing layer fields from source
+  const fieldsFromLayer = keptLayerFields.map(layerField => ({
+    // Default field properties from source
+    ...sourceFields.find(({ field }) =>
+      layerField.field === field),
+    // Field properties from layer
+    ...layerField,
+  }));
+
+
+  // All fields from source that are not in layerFields
+  const fieldsFromSource = sourceFields.filter(
+    ({ field }) =>
+      !layerFields.some(layerField =>
+        field === layerField.field),
+  );
+
+  return [
+    ...fieldsFromLayer,
+    ...fieldsFromSource,
+  ];
+};
+
+export const updateFieldFromSource = async (layerFields, dispatch, dataProvider, sourceId) => {
+  const { data: { fields: sourceFields = [] } } =
+      await dataProvider(GET_ONE, RES_DATASOURCE, { id: sourceId });
+
+  const sourceFieldMapped = sourceFields.map(({ id, ...sourceField }) =>
+    ({ field: id, sourceFieldId: id, ...sourceField }));
+
+  const result = mergeFields(layerFields || [], sourceFieldMapped);
```

Comment:
To avoid this kind of trick `|| []`, maybe you can initialise default value in the function line 25 ⬇️ 
`const mergeFields = (layerFields = [], sourceFields = []) => {`

Reply:
And so, in this function you can make an early return
`if(!layerFields.length && !sourceFields.length) return []`

Reply:
Done for first comment but for the second comment this case is so rare (should never happens) that i don't want an useless test.